File: runtests/python_lua/lua_require.py
from __future__ import annotations
from dataclasses import dataclass
from .precedences import Operator, binop_reduce
from lark import Token
import typing 

if typing.TYPE_CHECKING:
    from .lua_construct import expr
    op = typing.Union[Operator, expr]
else:
    op = object

def mkOperator(token: Token):
    opstr = str(token)
    return Operator(opstr, token)

def mkOperand(e: expr):
    return e

def mkBinOpSeq(xs: typing.List[op], make, _) -> expr:
    from .lua_construct import Bin
    return binop_reduce(lambda op: lambda l, r: make(op.val, l, r), xs)

def appendList(xs, a):
    xs.append(a)
    return xs    

def listMap(xs, f):
    return list(map(f, xs))

def some(a):
    return a

def none():
    return None


if typing.TYPE_CHECKING:
    maybe = typing.Optional
else:
    maybe = object

File: runtests/ocaml_lua/Lua_require.ml
open Lua_lexer;;
let appendList(xs, x) = xs @ [x]
let listMap(xs, f) = List.map f xs

type 'a maybe = 'a option

let some : 'a . 'a -> 'a maybe = fun a -> Some a
let none : 'a . unit -> 'a maybe =
    fun () -> None

type 'a op =
    | Operator of tbnf_token
    | Operand of 'a

let mkBinOpSeq = fun (xs, _, f) -> f xs

let mkOperator x = Operator x
let mkOperand x = Operand x